File: src/main/golang/leetcode/editor/cn/[105]Construct Binary Tree from Preorder and Inorder Traversal.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	var j int
	var build func(l, r int) *TreeNode
	build = func(l, r int) *TreeNode {
		if l > r {
			return nil
		}
		curr := preorder[j]
		var idx = -1
		for i := 0; i <= r; i++ {
			if inorder[i] == curr {
				idx = i
				break
			}
		}
		node := &TreeNode{Val: curr}
		j++
		node.Left, node.Right = build(l, idx-1), build(idx+1, r)
		return node
	}
	return build(0, len(inorder)-1)
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[105]Construct Binary Tree from Preorder and Inorder Traversal.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root = TreeNode(preorder[0])
        root_index = inorder.index(preorder[0])

        root.left = self.buildTree(preorder[1:root_index + 1], inorder[:root_index])
        root.right = self.buildTree(preorder[root_index + 1:], inorder[root_index + 1:])

        return root
# leetcode submit region end(Prohibit modification and deletion)
